Services de remplissage de base de données à partir d'un fichier Json

Morgan-sama opened the pull request.

Morgan-sama committed:
- Services de remplissage de base de données à partir d'un fichier Json

- Modification du docker compose pour une visualisation claire de la base de données conteneurisée + Regroupement de mysql et adminer (service) dans un même réseau ("appli")

- Changement du port d'execution du serveur d'application (8080 => 8081)
- Petits changements dans la classe Fields (Peut-être temporaires)

Juknum commented on backend/src/main/java/com/filegenie/backend/Controllers/RegisterFieldController.java:
```diff
@@ -0,0 +1,66 @@
+    @Autowired
+    private UserService userService;
+
+    @PostMapping(path = "/newFields")
```
```suggestion
    @PostMapping(path = "/")
```

Pas besoin de dire "newFields" si on est déjà sur l'URL `/api/file/import` et en POST

Juknum commented on backend/src/main/java/com/filegenie/backend/Controllers/RegisterFieldController.java:
```diff
@@ -0,0 +1,66 @@
+
+@Controller
+@RestController
+@RequestMapping("/api/register")
+public class RegisterFieldController {
```
```suggestion
@RequestMapping("/api/file/import")
public class ImportFileController {
```

(faudrait aussi renommer le fichier `.java` dans ce cas)

Juknum commented on backend/src/main/java/com/filegenie/backend/Controllers/RegisterFieldController.java:
```diff
@@ -0,0 +1,66 @@
+
+    @PostMapping(path = "/newFields")
+    public ResponseEntity<HttpStatus> AddNewField(
+            @RequestParam("file") MultipartFile jsonFile,
```
```suggestion
            @RequestParam("file") MultipartFile file,
```
Vu qu'on sait pas encore s'il s'agit d'un fichier JSON ou n'importe quoi d'autre

Morgan-sama replied:
Oui , intéressant je vais changer ça merci !

Juknum commented on backend/src/main/java/com/filegenie/backend/Controllers/RegisterFieldController.java:
```diff
@@ -0,0 +1,66 @@
+            @RequestParam("file") MultipartFile jsonFile,
+            @RequestHeader("Authorization") String token) {
+
+        User user;
+        try {
+            user = userService.getAuthedUser(token);
+        } catch (Exception e) {
+            String customMessage = "Token introuvable  : " + e.getMessage();
+            System.err.println(customMessage);
+            e.printStackTrace();
+            return new ResponseEntity<>(HttpStatus.UNAUTHORIZED);
+        }
```
```suggestion
        User user;

        try {
            user = userService.getAuthedUser(token);
        } catch (HttpException notFound) {
            return new ResponseEntity<>(notFound, notFound.getStatus());
        }
```

Tout pareil que dans `UserController`

Morgan-sama replied:
Juste pour tester à la base , mais j'ai oublié finalement de le retirer sinon c'est effectivement dans `UserController` qui j'ai pris ce bout de code

Juknum commented on backend/src/main/java/com/filegenie/backend/Controllers/RegisterFieldController.java:
```diff
@@ -0,0 +1,66 @@
+            return new ResponseEntity<>(HttpStatus.UNAUTHORIZED);
+        }
+
+        try {
+
+            File tempFile = File.createTempFile("uploaded-", ".json");
+            jsonFile.transferTo(tempFile);
+            List<Field> list = jsonParserService.parseJsonToFields(tempFile);
+            System.out.println(list);
+            registerFieldService.saveFields(list);
+            tempFile.delete();
+
+        } catch (Exception e) {
+            String customMessage = "Erreur lors de l'enregistrement des champs : " + e.getMessage();
+            System.err.println(customMessage);
+            e.printStackTrace();
+            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+        }
```
Ici faudrait séparer la reconnaissance du fichier (que ça soit JSON, XML, etc.) et appeler le service correspondant (que JSON pour le moment, mais prévoir avec un switch? en fonction du [type mime](https://developer.mozilla.org/en-US/docs/Web/HTTP/MIME_types) du fichier)

Un truc genre:

```java
File tmpFile = File.createTempFile("uploaded-", ".json");
file.transferTo(tmpFile);

// on recup le type mime
String mime = // TODO
// on prévoit le stockage des champs
List<Field> fields = new ArrayList<Field>();

switch (mime) {
  case "json":
  try {
    fields = jsonParserService.parseJsonToFields(tmpFile);
  } catch (Exception e) { // A voir si on peut mettre une Exception + précise
    return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
  }
  break;

  default:
    return new ResponseEntity<>(/* code HTTP pour un fichier non supporté? */);
}

if (/* fields à des items */) {
  registerFieldService.saveFields(fields);
}

return new ResponseEntity<>(HttpStatus.CREATED);

```

Morgan-sama replied:
Oui évidemment je vais ajuster le code une fois que j'aurai complètement fini avec le Json , histoire de ne pas m'embrouiller

Juknum commented on backend/src/main/java/com/filegenie/backend/Services/UserService.java:
```diff
@@ -59,6 +59,7 @@ public String loginUser(LoginRequest req) throws HttpException {
             session.setExpiresAt(LocalDateTime.now().plusHours(2));
 
             userSessionRepository.save(session);
+            System.out.println("Voici le token de l'utilisateur : "+ session.getSessionToken());
```
On peut l'enlever ?

```suggestion
```

Morgan-sama replied:
Oui , c'était juste pour faire mes test , à terme on va l'enlever tqt

Juknum reviewed (changes requested):
Ça à l'air pas mal gg

Morgan-sama committed:
Changement du port du backend dans le fichier fetch.ts (8080 => 8081)
Simplification du code (Renommage du controller d'import de fichier, changement des routes, et suppression de code inutile)

Juknum commented on backend/src/main/java/com/filegenie/backend/Controllers/ImportFileController.java:
```diff
@@ -0,0 +1,49 @@
+            registerFieldService.saveFields(list);
+            tempFile.delete();
+
+        } catch (Exception e) {
+            String customMessage = "Erreur lors de l'enregistrement des champs : " + e.getMessage();
+            System.err.println(customMessage);
+            e.printStackTrace();
+            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+        }
```
Du coup là faudrait que ça suive le format qu'on retrouve ailleurs: 

```suggestion
        } catch (HttpException httpError) {
                return new ResponseEntity<>(httpError, httpError.getStatus());
        }
```

Morgan-sama replied:
En changeant j'ai des erreurs de syntaxes qui me disent qu'en je dois gérer les erreurs de `file.transferTo(tempFile)` et File `tempFile = File.createTempFile("uploaded-", ".json")` et en utilisant la solution proposée par l'IDE ça me rajoute deux autres `catch` sur celui que j'ai déjà , donc je vais d'abord laisser tel quel.

Juknum commented on backend/src/main/java/com/filegenie/backend/Services/JsonParserService.java:
```diff
@@ -0,0 +1,40 @@
+@Service
+public class JsonParserService {
+
+    public List<Field> parseJsonToFields(File jsonFile) throws Exception {
+        ObjectMapper objectMapper = new ObjectMapper();
+        JsonNode rootNode = objectMapper.readTree(jsonFile);
+        return processNode(null, rootNode);
```
Et là aussi pour que ça colle avec le controller:
```suggestion
    public List<Field> parseJsonToFields(File jsonFile) throws HttpException
        try {
            ObjectMapper objectMapper = new ObjectMapper();
            JsonNode rootNode = objectMapper.readTree(jsonFile);
            return processNode(null, rootNode);
        } catch (Exception e) {
            throw new HttpException(HttpStatus.INTERNAL_SERVER_ERROR, "Impossible de parser le fichier JSON: " + e.getMessage());
        }
```

Juknum committed:
fix: remove left println

Signed-off-by: Julien Constant <julienconstant190@gmail.com>

Juknum reviewed (changes requested):
Plus que ça et ça devrait être gulli good mdr

Morgan-sama committed:
Ajustements dans JsonParserService.java (Ajout d'une vérification des fichiers et gestion des erreurs)

Morgan-sama commented on backend/src/main/java/com/filegenie/backend/Services/UserService.java:
Mec , j'en ai toujours besoin pour l'associer à mes requêtes postman, vu que je ne passe pas par le frontend pour faire mes tests

Juknum replied:
AAh merde my bad, tu peux le choper dans la base de donnée sinon

Morgan-sama committed:
Modification de la classe "Values" => "FielValue" pour moins d'ambiguîté , Ajout d'une relation @oneTomany entre Field et FieldValue

Juknum commented:
Si c'est tout bon je te laisse merge du coup @Morgan-sama
